doc/cookbook/current-route/CurrentRoute.lhs

```diff
@@ -0,0 +1,391 @@
+        passToServer d Wai.rawPathInfo
+```
+
+Using new `RawPathInfo` combinator, we can re-implement our application:
+
+```haskell
+data Routes2 route = Routes2
+    { route21 :: route :- RawPathInfo :> "route1" :> Get '[JSON] Int
+    , route22 :: route :- RawPathInfo :> "route2" :> Capture "id" Int :> Get '[JSON] Int
+    }
+  deriving (Generic)
+
+routes2 :: Routes2 AsServer
+routes2 = Routes2
+    { route21 = \p -> liftIO $ do
+        BS8.putStrLn p
+        return (BS8.length p)
+    , route22 = \p i -> liftIO $ do
+        BS8.putStrLn p
+        return (succ i)
+    }
+
+app2 :: Application
+app2 = genericServe routes2
+```
+
+This version is slightly more declarative, and handlers are more restrictive.
+We moved the `rawPathInfo` selector from handlers to combinator implementation,
+removed repetition.
+
+### Using `Vault`
+
+The `vault` value in `wai` `Request` is not well known or used.
+But in this scenario it can be useful.
+Vault is explained in [Using WAI's vault for fun and profit](https://www.yesodweb.com/blog/2015/10/using-wais-vault) blog post.
+It fills a "dynamic" gap of strongly typed `Request`: we can attach arbitrary data to the request,
+as is common in web frameworks in a dynamically typed languages.
+As `servant-server` is based on `wai`, using vault is the third answer
+to the first part of the question.
+
+We (unsafely) create a key to the vault:
+
+```haskell
+rpiKey :: V.Key BS8.ByteString
+rpiKey = unsafePerformIO V.newKey
+```
+
+Then we create a *middleware* which will put `rawPathInfo` into the `vault`.
+
+```haskell
+middleware :: Wai.Middleware
+middleware app req respond = do
+    let vault' = V.insert rpiKey (Wai.rawPathInfo req) (Wai.vault req)
+        req' = req { Wai.vault = vault' }
+    app req' respond
+```
+
+Using this we make third variant of our application.
+Note that we values might not be in the vault,
+that's small functional regression.
+
+```haskell
+data Routes3 route = Routes3
+    { route31 :: route :- Vault :> "route1" :> Get '[JSON] Int
+    , route32 :: route :- Vault :> "route2" :> Capture "id" Int :> Get '[JSON] Int
+    }
+  deriving (Generic)
+
+routes3 :: Routes3 AsServer
+routes3 = Routes3
+    { route31 = \v -> liftIO $ do
+        let p = fromMaybe "?" $ V.lookup rpiKey v
+        BS8.putStrLn p
+        return (BS8.length p)
+    , route32 = \v i -> liftIO $ do
+        let p = fromMaybe "?" $ V.lookup rpiKey v
+        BS8.putStrLn p
+        return (succ i)
+    }
+
+app3 :: Application
+app3 = middleware $ genericServe routes3
+```
+
+Note: that `vault` can be used to pass information from middlewares to handlers
+and from handlers to middlewares. For example, the authentication can be done
+completely in the middleware, with a user information stored in the vault for
+handlers to use.
+
+## How to get current *route*?
+
+The second part of a question, is how to get current route.
+Something, we can get `route2/:id` out?
```

Comment:
I'd rephrase: "Is there something that could give us `/route2/:id` for our example from earlier?"